EntityProvider - Add new 'civi.entity.fields' event

colemanw opened the pull request:
Overview
----------------------------------------
Adds a new EFv2 replacement for the legacy DAO `fields_callback` functions. Marks the latter as deprecated.

colemanw committed:
EntityProvier - Ensure getFields() always goes through a single codepath

This makes EntityProvider::getFields the single access-point for field metadata,
allowing us to add a hook at that point.

colemanw committed:
EntityProvier - Add new 'civi.entity.fields' event.

This is the EFv2 replacement for the legacy DAO `fields_callback` functions

colemanw committed:
DAO - Add code comments indicating `fields_callback` is deprecated

totten commented:
I've been doing an `r-run` to try registering a field -- having trouble at the moment. I *think* it should work but isn't. Debugging a bit more -- and if I can't get it, I'll post the fuller example.

colemanw committed:
DAO_Base - Switch to using entityProvider

Using the entityProvider object improves consistency, and ensures that the new 'civi.entity.fields' event
is respected for field alterations.

totten commented:
> I started to port those updates to the `CiviMix\Schema` version of `DAO_Base`... but then... (_n.m._)

Yeah, that's fine IMHO. I don't think it's important to support `civi.entity.fields` on old versions. This means `civi.entity.fields` will be available on 6.7+.

Note that the `CiviMix\\Schema\\Xxxx\\DAO` is only relevant on sites running 5.74 or earlier. Once you're running on 5.74+, that polyfill isn't used. ([example](https://github.com/civicrm/civicrm-core/blob/14fec78a3a81eade22c12cf24fccd2ecaa6da84a/ext/oauth-client/oauth_client.civix.php#L95-L106)) It is humanly possible to rework that to backport this event, but... I don't really think we need to...

> There's no code comment to explain the difference, but I'm guessing it's avoiding `AllCoreTables` for a reason, and if so then we can't port these changes. That's also a can we could kick down the road without holding up this PR.

Close -- yes, it is purposeful, but I don't think it's specifically `AllCoreTables`. Rather, it's the second line:

```php
 private static function getEntityDefinition(): array { 
   $entityName = CRM_Core_DAO_AllCoreTables::getEntityNameForClass(static::class); 
   return \Civi\Schema\EntityRepository::getEntity($entityName); 
 } 
```

`AllCoreTables` goes way-back, and `civimix-schema@5/src/DAO.php` does call it multiple times. However, `EntityRepository` does the heavy lifting here - it's also newer and harder to backport.

Comparing the two snippets, I would summarize the polyfill version as saying:  "_We need to read `*.entityType.php`, but we don't have `EntityRepository`, so we'll just read the file directly._"

totten commented:
`r-run` looks good now. 🎉 

`r-explain` is a little thin, but @colemanw is planning to post a longer update for dev-docs.

Merge it!

colemanw commented:
Thanks @totten. Here's the promised docs: https://lab.civicrm.org/documentation/docs/dev/-/merge_requests/1236
